mcumgr: Bootloader info request definition (#57797)
zephyrproject-rtos/zephyr: doc/services/device_mgmt/smp_groups/smp_group_0.rst

    +-------------------+-----------------------------------------------+
    | ``7``             | OS/Application info                           |
    +-------------------+-----------------------------------------------+

>     | ``8``             | Bootloader parameters                         |

Reviewer: `Bootloader Information`

Reviewer 2: Seems unaddressed

Reviewer 2: I wonder if group 0 is the right place for this and if group 1 is more suited since it is related to images

Author: Fixed

    | "rc"         | :c:enum:`mcumgr_err_t`                        |
    |              | only appears if non-zero (error condition).   |
    +--------------+-----------------------------------------------+

Bootloader Information
**********************

Allows to obtain information on board bootloader and its parameters.

Bootloader Information Request
==============================

Bootloader information request header:

.. table::
    :align: center

    +--------+--------------+----------------+
    | ``OP`` | ``Group ID`` | ``Command ID`` |
    +========+==============+================+
    | ``0``  | ``0``        |  ``8``         |
    +--------+--------------+----------------+

CBOR data of request:

.. code-block:: none

    {
        (str,opt)"query"  : (str)
    }

where:

.. table::
    :align: center

    +--------------+-----------------------------------------------+

>     | "query"      | Is string representing query for parameters,  |

Reviewer 3: "A string representing query parameter. There is no restrictions on the value as processing of the query is left for the bootloader backend."

>     | "query"      | Is string representing query for parameters,  |

Reviewer 3: Ping

    |              | with no restrictions how the query looks like |
    |              | as processing of query is left for bootloader |
    |              | backend.                                      |
    +--------------+-----------------------------------------------+
    | "rc"         | :c:enum:`mcumgr_err_t`                        |
    |              | only appears if non-zero (error condition).   |

>     +--------------+-----------------------------------------------+

Author: Will fix.

Bootloader Information Response
===============================

Bootloader information response header:

.. table::
    :align: center

    +--------+--------------+----------------+
    | ``OP`` | ``Group ID`` | ``Command ID`` |
    +========+==============+================+

>     | ``2``  | ``0``        |  ``8``         |

Reviewer: btw. Shouldn't it be `1` - `MGMT_OP_READ_RSP`?

Reviewer 3: Yes, and btw for App Info (already merged) as well:
https://builds.zephyrproject.io/zephyr/pr/59345/docs/services/device_mgmt/smp_groups/smp_group_0.html#os-application-info-response

Reviewer 2: Unaddressed

Author: Done

In case when no "query" has been provided in request,
CBOR data of response:

.. code-block:: none

    {
        (str)"bootloader"      : (str)
        (opt,str)"rc"          : (int)
    }

where:

.. table::
    :align: center

    +--------------+-----------------------------------------------+
    | "bootloader" | String represengint bootloader name           |
    +--------------+-----------------------------------------------+
    | "rc"         | :c:enum:`mcumgr_err_t`                        |
    |              | only appears if non-zero (error condition).   |
    +--------------+-----------------------------------------------+

In case when "query" is provided:

.. code-block:: none

    {
        (str,opt)<param_name>   : ()
        ...
        (opt,str)"rc"           : (int)
    }

where:

.. table::
    :align: center

    +--------------+-----------------------------------------------+
    | <param_name> | Response to "query". This is optional and may |
    |              | be left out in case when query yields no      |
    |              | response, but the "rc" code needs to          |
    |              | `MGMT_ERR_ENOENT`.                            |

>     |              | Response may have more than one paraeter      |

Reviewer: parameter

Author: Done

    |              | reported back or it may be a map, that is     |
    |              | dependent on bootloader backednd and query.   |
    +--------------+-----------------------------------------------+

>     | "rc"         | :c:enum:`mcumgr_err_t`                        |

Reviewer: I wonder if we should return anything in response in case of error. Do you want to allow returning an incomplete response in that case?
Alternatively we may return only `rc` (on error) or only response (on success)

Reviewer 2: Should use the new group error codes rather than rc

    +--------------+-----------------------------------------------+

Bootloader Information: MCUboot
===============================

In case when MCUboot is application bootloader empty request will
be responeded with:

.. code-block:: none

    {
        (str)"bootloader"      : (str)"MCUboot"
    }

Following parameter requests are supported:

.. code-block:: none

    {
	(str)"mode"		: (int)<anything>
    }

where `<anything>` is any integer number which, in this case, is ignored.

Response to "mode" is:

.. code-block:: none

    {
	(str)"mode"		: (int)
    }

where "mode" is one of:

.. table::
    :align: center

    +-----+-------------------------------------------------+
    | -1  | Unknown mode of MCUboot			    |
    +-----+-------------------------------------------------+
    |  0  | MCUboot is in single application mode           |
    +-----+-------------------------------------------------+
    |  1  | MCUboot is in swap using scratch partition mode |
    +-----+-------------------------------------------------+
    |  2  | MCUboot is in swap without scratch mode         |
    +-----+-------------------------------------------------+

>     |  3  | MCUboot is in DirectXIP mode                    |

Author: Probably, currently we do not have Kconfig options for them anyway.

>     |  3  | MCUboot is in DirectXIP mode                    |

Reviewer 3: DirectXIP without revert is missing.

>     |  3  | MCUboot is in DirectXIP mode                    |

Reviewer 3: OK, so 3 is without revert and 4 is with revert, perhaps wise to add "without revert" in 3?

    |              | `MGMT_ERR_ENOENT` shall be returned when      |
    |              | query has no result.                          |
    +--------------+-----------------------------------------------+

Bootloader Information: MCUboot
===============================

In case when MCUboot is application bootloader empty request will
be responded with:

.. code-block:: none

    {
        (str)"bootloader"      : (str)"MCUboot"
    }

Currently "MCUboot" supports querying for mode of operation:

.. code-block:: none

    {
        (str)"query"           : (str)"mode"
    }

Response to "mode" is:

.. code-block:: none

    {

>         (str)"mode"            : (int)

Reviewer: Should we inform that `rc` should be part of the response? Dismiss this comment if you want to define separate format for successful (just strings) and error (just rc) response.

    | "rc"             | :c:enum:`mcumgr_err_t` only appears if non-zero (error condition) when  |
    |                  | using SMP version 1 or for SMP errors when using SMP version 2.         |
    +------------------+-------------------------------------------------------------------------+

Bootloader Information
**********************

> Allows to obtain information on board bootloader and its parameters.

Reviewer 2: ```suggestion
Allows retrieving information about the on-board bootloader and its parameters.
```

    +------------------+-------------------------------------------------------------------------+
    | "err" -> "rc"    | contains the index of the group-based error code. Only appears if       |
    |                  | non-zero (error condition) when using SMP version 2.                    |
    +------------------+-------------------------------------------------------------------------+
    | "rc"             | :c:enum:`mcumgr_err_t` only appears if non-zero (error condition) when  |
    |                  | using SMP version 1 or for SMP errors when using SMP version 2.         |
    +------------------+-------------------------------------------------------------------------+

Bootloader Information: MCUboot
===============================

In case when MCUboot is application bootloader empty request will
be responded with:

.. code-block:: none

    {
        (str)"bootloader"      : (str)"MCUboot"
    }

Currently "MCUboot" supports querying for mode of operation:

.. code-block:: none

    {
        (str)"query"           : (str)"mode"
    }

Response to "mode" is:

.. code-block:: none

    {
        (str)"mode"             : (int)

>         (str,opt)"no-dgrade"    : (bool)

Reviewer 2: would use full word here, "no-downgrade"

    }

where "mode" is one of:

.. table::
    :align: center

    +-----+-------------------------------------------------+
    | -1  | Unknown mode of MCUboot                         |
    +-----+-------------------------------------------------+
    |  0  | MCUboot is in single application mode           |
    +-----+-------------------------------------------------+
    |  1  | MCUboot is in swap using scratch partition mode |
    +-----+-------------------------------------------------+
    |  2  | MCUboot is in swap without scratch mode         |
    +-----+-------------------------------------------------+
    |  3  | MCUboot is in DirectXIP without revert mode.    |
    +-----+-------------------------------------------------+
    |  4  | MCUboot is in DirectXIP with revert mode        |

>     +-----+-------------------------------------------------+

Reviewer 2: RAM loader seems to be missing

    +--------------+-----------------------------------------------+

Bootloader Information Response
===============================

Bootloader information response header:

.. table::
    :align: center

    +--------+--------------+----------------+
    | ``OP`` | ``Group ID`` | ``Command ID`` |
    +========+==============+================+
    | ``1``  | ``0``        |  ``8``         |
    +--------+--------------+----------------+

In case when no "query" has been provided in request,
CBOR data of response:

.. code-block:: none

    {
        (str)"bootloader"      : (str)
    }

where:

.. table::
    :align: center

    +--------------+-----------------------------------------------+
    | "bootloader" | String representing bootloader name           |
    +--------------+-----------------------------------------------+

In case when "query" is provided:

.. code-block:: none

    {
        (str,opt)<param_name>   : ()
	...
    }

where:

.. table::
    :align: center

    +------------------+-------------------------------------------------------------------------+
    | <param_name>     | Response to "query". This is optional and may be left out in case when  |

>     |                  | query yields no response, but the "rc" code needs to `MGMT_ERR_ENOENT`. |

Reviewer 2: Don't think this error code is right, it's `OS_MGMT_ERR_QUERY_YIELDS_NO_ANSWER`?

    +------------------+-------------------------------------------------------------------------+
    | "rc"             | :c:enum:`mcumgr_err_t` only appears if non-zero (error condition) when  |
    |                  | using SMP version 1 or for SMP errors when using SMP version 2.         |
    +------------------+-------------------------------------------------------------------------+

Bootloader Information: MCUboot
===============================

In case when MCUboot is application bootloader empty request will
be responded with:

.. code-block:: none

    {
        (str)"bootloader"      : (str)"MCUboot"
    }

Currently "MCUboot" supports querying for mode of operation:

.. code-block:: none

    {
        (str)"query"           : (str)"mode"
    }

Response to "mode" is:

.. code-block:: none

    {
        (str)"mode"                     : (int)
        (str,opt)"no-downgrade"         : (bool)
    }

where "mode" is one of:

.. table::
    :align: center

    +-----+-----------------------------------------------------+
    | -1  | Unknown mode of MCUboot.                            |
    +-----+-----------------------------------------------------+

>     |  0  | MCUboot is in single application mode.              |

Reviewer 2: Probably best to use these enum values for this: https://github.com/mcu-tools/mcuboot/blob/main/boot/bootutil/include/bootutil/boot_status.h#L118

Author: Done

    +-----+-----------------------------------------------------+
    |  1  | MCUboot is in swap using scratch partition mode.    |
    +-----+-----------------------------------------------------+
    |  2  | MCUboot is in swap without scratch mode.            |
    +-----+-----------------------------------------------------+
    |  3  | MCUboot is in DirectXIP without revert mode.        |
    +-----+-----------------------------------------------------+
    |  4  | MCUboot is in DirectXIP with revert mode.           |
    +-----+-----------------------------------------------------+
    |  5  | MCUboot is in RAM loader mode.                      |
    +-----+-----------------------------------------------------+

> The "no-dgrade" is a flag, which is always sent when true, indicating that

Reviewer 2: *no-downgrade

===============================

Bootloader information response header:

.. table::
    :align: center

    +--------+--------------+----------------+
    | ``OP`` | ``Group ID`` | ``Command ID`` |
    +========+==============+================+
    | ``1``  | ``0``        |  ``8``         |
    +--------+--------------+----------------+

In case when no "query" has been provided in request,
CBOR data of response:

.. code-block:: none

    {
        (str)"bootloader"      : (str)
    }

where:

.. table::
    :align: center

    +--------------+-----------------------------------------------+
    | "bootloader" | String representing bootloader name           |
    +--------------+-----------------------------------------------+

In case when "query" is provided:

.. code-block:: none

    {
        (str,opt)<param_name>   : ()
	...
    }

where:

.. table::
    :align: center

    +------------------+-------------------------------------------------------------------------+
    | <param_name>     | Response to "query". This is optional and may be left out in case when  |
    |                  | query yields no response, but the "rc" code needs to                    |

>     |                  | `OS_MGMT_ERR_QUERY_YIELDS_NO_ANSWER`.                                   |

Reviewer 2: Would put something like "SMP version 2 error code of..."

Author: Done

    |                  | Response may have more than one parameter reported back or it may be    |

>     |                  | a map, that is dependent on bootloader backednd and query.              |

Reviewer 3: Typo: backend

    +-----+-----------------------------------------------------+
    |  0  | MCUboot is in single application mode.              |
    +-----+-----------------------------------------------------+
    |  1  | MCUboot is in swap using scratch partition mode.    |
    +-----+-----------------------------------------------------+
    |  2  | MCUboot is in overwrite (upgrade-only) mode.        |
    +-----+-----------------------------------------------------+
    |  3  | MCUboot is in swap without scratch mode.            |
    +-----+-----------------------------------------------------+
    |  4  | MCUboot is in DirectXIP without revert mode.        |
    +-----+-----------------------------------------------------+
    |  5  | MCUboot is in DirectXIP with revert mode.           |
    +-----+-----------------------------------------------------+
    |  6  | MCUboot is in RAM loader mode.                      |
    +-----+-----------------------------------------------------+

The "no-downgrade" is a flag, which is always sent when true, indicating that
mode has downgrade prevention enabled; downgrade prevention means that

> if uploaded image has lower version than running, it will notbe taken

Reviewer 3: not be*

> for exectuion by MCUboot.

Reviewer 3: execution
